version: "3.8"
services:
  client:
    container_name: wt_client
    build:
      context: ./client
      dockerfile:
        Dockerfile.prod
    restart: on-failure
    stdin_open: true
    volumes:
      - ./client:/app
    environment:
      VIRTUAL_HOST: weight-tracker.zirr.dev
      LETSENCRYPT_HOST: weight-tracker.zirr.dev
      VITE_BASE_API_URL: api.weight-tracker.dev
    expose:
      - 80

  api:
    container_name: wt_api
    build:
      context: ./api
      dockerfile:
        Dockerfile.prod
    restart: on-failure
    depends_on:
      - mongodb
    links:
      - mongodb
    ports:
      - "3001:3001"
    volumes:
      - ./api:/app
      - /app/node_modules
    environment:
      VIRTUAL_HOST: api.weight-tracker.zirr.dev
      LETSENCRYPT_HOST: api.weight-tracker.zirr.dev

  mongodb:
    container_name: mongodb
    image: mongo:6.0
    volumes:
        - 'dbdata:/data/db'
    ports:
      - "27017:27017"
    healthcheck:
      interval: 10s
      retries: 300
      test: ["CMD", "mongo", "--quiet", "localhost", "--eval", "db.runCommand({ ping: 1 }).ok"]
      timeout: 10s

  nginx-proxy:
    image: jwilder/nginx-proxy
    container_name: nginx-proxy
    ports:
      - "80:80"
      - "443:443"
    volumes:
      - conf:/etc/nginx/conf.d
      - vhost:/etc/nginx/vhost.d
      - html:/usr/share/nginx/html
      - dhparam:/etc/nginx/dhparam
      - certs:/etc/nginx/certs:ro
      - /var/run/docker.sock:/tmp/docker.sock:ro

  letsencrypt-nginx-proxy-companion:
    image: jrcs/letsencrypt-nginx-proxy-companion
    container_name: nginx-proxy-letsencrypt
    volumes_from:
      - nginx-proxy
    volumes:
      - certs:/etc/nginx/certs:rw
      - /var/run/docker.sock:/var/run/docker.sock:ro
    network_mode: bridge

networks:
  default:
    external:
      name: nginx-proxy

volumes:
  dbdata:
  certs:
  vhost:
  conf:
  html:
  dhparam:
